NuoDB Backup and Restore in Kubernetes (doc)

sivanov-nuodb opened the pull request:
Added user guide on how to use Backup and Restore mechanics with NuoDB Helm Charts in Kubernetes.

vegichan commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+
+After database installation, the available backup and restore features supported by these mechanisms are: 
+
+1. [Automatic archive initial import](#automatic-archive-initial-import)
```
is the initial import done "after database installation"?

sivanov-nuodb replied:
It is done after `database` chart installation by `nuosm` before SMs database processes are started.

vegichan commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+### Scheduled online database backups
+
+Multiple backup cron jobs are scheduled as part of the [database](../stable/database) chart deployment by default.
+The backup schedules can be customized to meetdifferent backup and recovery objectives.
```
missing space between meet, different

sivanov-nuodb replied:
Fixed

vegichan commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+The archive seed restore operation restores a corrupted or lost archive while the database is running.
+This operation will reset the existing archive state, and once it joins the database, the new SM will sync to the current state of the running database.
+For this reason, you cannot restore the entire database by restoring a single SM in a running database.
+The _seed restore_ reduces the _SYNCing_ time from other running SMs as only the changed atoms will be transferred.
```
in general, that's exactly what SYNC does too. In the absence of bugs, a full SM sync should be as fast as seed+diff-sync. I know that in reality, it is not.

But can we still change the wording to something like "The seed restore - has the potential to reduce - the syncing time"?

sivanov-nuodb replied:
I see the _archive seed restore_ only used when the archive needs to be re-initialized. It compares full archive sync with seed+diff sync here. In absence of bugs, the seed+diff sync should be a lot faster than transferring all atoms.
I agree that in any other use case it's hard to know what the gain will be so I will change the text as suggested.

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+Each _full_ hot copy will create a backup set with the same name in the backup root directory (by default `/var/opt/nuodb/backup`) on all HC SMs in a single database and a single backup group.
+By default, all HC SMs in a single Kubernetes cluster are part of one backup group.
+This can be adjusted by changing the `database.hotcopy.backupGroup` value for multiple Kubernetes clusters in multi-cluster deployment.
+_Incremental_ and _journal_ hot copies are always stored in the current backup set if one exists.
```
The current backup set is always the latest one?

sivanov-nuodb replied:
The _current_ backup set is always the one that was issued (it may or may not have finished).

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+_Incremental_ and _journal_ hot copies are always stored in the current backup set if one exists.
+If there is no current backup set for some of the archives served by a HC SM when _incremental_ or _journal_ backup is requested, a _full_ hot copy is triggered automatically.
+The `nuobackup` script is made available as a configMap and is used by all backup cron jobs to send a hot copy request of the specified type to the HC SMs.
+The _Full_ hot copy result is recorded into the NuoDB Admin domain key-value (KV) store upon successful finish.
```
This seems to answer my questions above, but perhaps this could be reworded to make it clearer. Also, should we include the KV key format, or `nuocmd` invocation that can be used to obtain the current/latest full backup?

sivanov-nuodb replied:
Good point! I would suggest including the command wrappers rather than the KV prefixes.
The current backup set is available via `nuodocker get current backup --db-name <db>`.
The latest successful backup set is available via `nuobackup --type report-latest --db-name <db> --group <backup group>`.

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+The restore/import source location can be one of the following:
+- remote _URL_ in a form of `protocol://authority/path`.
+The restore logic uses [cURL](https://curl.se/) to download a stream of data into an SM pod and restore the archive from it.
+Standard protocols such as _HTTP/s_, _FTP_ and _SFTP_ can be used as a source.
```
I guess it can be any protocol that curl supports, since the URL is passed through verbatim. Or do we actually restrict the protocol to these three?

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+The restore logic uses [cURL](https://curl.se/) to download a stream of data into an SM pod and restore the archive from it.
+Standard protocols such as _HTTP/s_, _FTP_ and _SFTP_ can be used as a source.
+Any other protocols that have built-in support in the shipped `cURL` binary in the NuoDB image can be used as well.
+Only basic authentication can be specified for remote sources. A remote location that has an authentication token embedded into the _URL_ can also be used.
```
The second sentence sort of contradicts the "Only" part of the first. I would reword this as

> Basic authentication or an authentication token that is embedded in the URL are supported.

I guess this rules out certificate authentication or authentication other than basic auth that is passed as a request header (e.g. `Authorization: Bearer <token>`).

sivanov-nuodb replied:
Thanks! I will change it as suggested.

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+
+Two restore/import source types are available:
+- _stream_ - Used as a synonym for a copy of NuoDB archive which contains both the archive and journal directories.
+A remote _stream_ source is downloaded and extracted directly into the archive directory which doesn't require extra space for holding the restore/import source file.
```
Somewhat unrelated, but do we support separate journal directories at all with the StatefulSets?

sivanov-nuodb replied:
Currently no.

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+1. Identify the restore source which can be either a backup set created by scheduled backup jobs or a remote source.
+2. Identify which archives need to be restored.
+By default, all archives served by HC SMs in a single Kubernetes cluster will be selected for a restore as they can define the database state by restoring from a backup set available in the backup volumes.
+3. Ensure that the restore source is available on all SMs selected for restore.
```
I would add a note saying that if the restore source is a backup set, then only HC SMs will have access to the restore source.

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+2. Identify which archives need to be restored.
+By default, all archives served by HC SMs in a single Kubernetes cluster will be selected for a restore as they can define the database state by restoring from a backup set available in the backup volumes.
+3. Ensure that the restore source is available on all SMs selected for restore.
+4. Ensure that there is enough free disk space in the archive volume of each SM so that a copy of the archive before the restore is taken.
```
Is a copy taken, or are the existing contents of the archive moved? And what would the contents of the archive directory be after restore is complete for a restore and non-restore (i.e. syncing) SM. I would guess the following:
- Restore SM: Archive restored from backup set or copy ("stream"). The backup set and the archive directories would be on different volumes. I think for "stream" we are extracting the backup as we download it (e.g. `curl ... | tar -xf`). So in both cases, the archive volume would need enough space to accommodate the archive we want to restore to and the backup we create of the current contents.
- Non-restore SM: Wouldn't we need the archive to accommodate the backup we are creating and the archive that we would sync after start-up, which in the absence of user-defined SGs, would be the same as the "restore SM".

sivanov-nuodb replied:
Archive backup is taken only on SMs selected for restore.
The existing archive content is moved to `<db>-save-<timestamp>` and `nuodocker restore archive` restore into the original archive directory which is empty.
In case remote URL pointing to backup set is used as restore source, it will be downloaded into archive directory temporarily.
I will add these in the document.

acabrele commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+
+The table below shows the compatibility matrix for different backup and restore features in combination with the NuoDB product version and NuoDB Helm Charts.
+
+|                                       | NuoDB 4.0.x+ Helm Charts 2.x.x | NuoDB 4.0.x+ Helm Charts 3.0.x | NuoDB 4.0.x+ Helm Charts 3.1.x | NuoDB 4.2.x+ Helm Charts 3.2.x+ |
```
Can we make it clear somewhere that this documents functionality of the last column, and there is no previous equvilent document?

sivanov-nuodb replied:
I would say that some parts of the document are specific to the last column but not all. For example, the _Scheduled online database backups_, _Automatic archive initial import_, and _Automatic archive restore_ haven't been changed a lot since the original implementation.
This compatibility matrix was my attempt to show what is supported where and to point major changes in the B&R logic.
I agree that the examples and the sample output is specific to the latest column so I will call that out.

acabrele commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+
+The backup jobs execution should be monitored on a regular basis to ensure that they complete successfully.
+By default, each backup job execution will be retried on failure which can be configured by the  `database.hotCopy.restartPolicy` setting.
+There are no backup retention policies currently available. The admin user is responsible to make sure that there is enough space in the backup volume of each SM for new hot copies to be created.
```
This is really important info an feels a little lost between the other two sentences about the job histories. Can we pull it into a section on retention management?  Where we could also mentioned the possibility of using customer provided read write many volumes (EFS, azure file) enabling retention management more easily using cloud tooling for example.

sivanov-nuodb replied:
Good call! I don't know a lot about the "cloud tooling" mentioned here.
I will include the statement as is in a new section. If you would like me to include more info, please let me know.

acabrele commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+Standard protocols such as _HTTP/s_, _FTP_ and _SFTP_ can be used as a source.
+Any other protocols that have built-in support in the shipped `cURL` binary in the NuoDB image can be used as well.
+Only basic authentication can be specified for remote sources. A remote location that has an authentication token embedded into the _URL_ can also be used.
+The remote URL should point to the `tar.gz` file which will be automatically extracted using the configured `stripLevels` setting (by default `1`). 
```
Can we add what the aim is (e.g strip levels needs to remove all parent folders)

acabrele replied:
Also is this referring to the restore chart or the autoimport, or both (I thought striplevels was removed from the restore chart)

sivanov-nuodb replied:
> Can we add what the aim is (e.g strip levels needs to remove all parent folders)

The very next sentence is explaining what `stripLevels` are used for. Do you want me to rephrase it?

> I thought striplevels was removed from the restore chart

They have never been exposed in the `restore` chart which is a bug that we should fix.

acabrele replied:
Yes I see "These are the number of leading path elements removed by the tar command during archive extraction." But that doesn't say what the aim is or help someone know what it should be set to if it were non-default (should all leading elements be removed... should one parent be left etc)

sivanov-nuodb replied:
Ah, I see. Okay, I added some extra explanation.

acabrele commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+- `:group-latest` - The latest backup set name for the backup group in current database deployment.
+
+Two restore/import source types are available:
+- _stream_ - Used as a synonym for a copy of NuoDB archive which contains both the archive and journal directories.
```
Use the terminology you previously defined? - "Offline backup"

sivanov-nuodb replied:
Good point. I will change it.

acabrele commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+Two restore/import source types are available:
+- _stream_ - Used as a synonym for a copy of NuoDB archive which contains both the archive and journal directories.
+A remote _stream_ source is downloaded and extracted directly into the archive directory which doesn't require extra space for holding the restore/import source file.
+- _backupset_ - Hot copy backup set which is either fetched from a remote source or available in the backup directory.
```
like the previous line, explain this requires additional space for download/extract?

acabrele commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+1. Identify the restore source which can be either a backup set created by scheduled backup jobs or a remote source.
+2. Identify which archives need to be restored.
+By default, all archives served by HC SMs in a single Kubernetes cluster will be selected for a restore as they can define the database state by restoring from a backup set available in the backup volumes.
+3. Ensure that the restore source is available on all SMs selected for restore.
```
I thought we had changes so that if the source didn't exist it would proceed to sync instead?  Maybe we are more explicit now...

acabrele replied:
I guess there is a step/action here to copy the backup set to the SMs, not just check?  Which may not be possible if they are crash looping for example. 

We can't guarantee that a backupset name will be identical on all HCSMs just because they all ran on the same schedule,there could be seconds different

sivanov-nuodb replied:
With the latest changes, all SMs selected for a restore should be restored so that database restore is performed.
By default, only HC SMs are selected to be restored. The rest of the SMs will sync as before.

I saw aiming to consolidate all the restore source rangling in step 3. As you pointed out, the desired backup set could be made available as a remote URL or copied in the backup volume manually, etc.

> We can't guarantee that a backupset name will be identical on all HCSMs just because they all ran on the same schedule

`nuodocker backup database` guarantees that the backup set names are identical for all selected archives in each _full_ hot copy operation.

acabrele commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+5. Invoke the database restore request by installing NuoDB _restore_ chart.
+By default, the database set as `restore.target` will be **shutdown** and restarted.
+
+Restoring to a previous point in time starts by installing the `restore` chart which will invoke the database restore request.
```
Careful with terminology overlap with true PiT restore

sivanov-nuodb replied:
Oops ... you are right.

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+Restoring to a previous point in time starts by installing the `restore` chart which will invoke the database restore request.
+In this example, a specific backup set that is available on all HC SMs will be used as a restore source.
+
+> **NOTE**: If the restore source is set to `:latest` or `:group-latest`, it's recommended that the incremental and journal backups are temporarily suspended. This is to ensure that another backup is not accidentally taken.
```
Would it be possible to check the "restore request" metadata in `nuobackup` and would that solve this race condition?

sivanov-nuodb replied:
Yes, the restore source is available in the restore request. 
It's more of a question, should `nuobackup` abort the requested backup if a database restore using `:latest` or `:group-latest` is requested?

acabrele commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+
+1. Delete the database chart deployment.
+2. Remove the associated PVC, PV, and cloud storage volume that remains after database deletion.
+3. Remove the database from the domain by using `nuocmd delete database --db-name <db>`.
```
Describe the option of deleting the nuoadmin storage too? (instead of nuocmd delete)

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+The same applies if the reason for the previously failed database restore is corrected.
+When troubleshooting distributed database restore failures, it's important to collect logs from all database processes and their corresponding pods.
+
+> **NOTE**: In multi-cluster deployments, different backup cron jobs execute with different schedules unless they are configured to be part of the same backup group.
+
+Backup sets produced by different backup groups can't be used together to perform _distributed database in-place restore_.
+To perform a database restore in multi-cluster deployment, proceed with one of the approaches described below.
+
+1. [Select the hot copy SMs in one of the clusters](#fine-grained-archive-selection) and provide a backup set available in that cluster or `:group-latest` as `restore.source`.
+2. Make the selected backup set available as a remote URL in both clusters. It can be used as a restore source for all the SMs in the database only if no custom storage groups are configured.
+3. Configure HC SMs in all clusters to be part of a single backup group and leave one set of backup jobs enabled in only one of the clusters to control this backup group.
+This will ensure that all HC SMs backups will be coordinated during hot copy requests.
+
```
This seems kind of awkward and kind of defeating the purpose of the HC SM StatefulSet. The user has to manually choose a specific backup set to make available remotely, presumably from among the HC SMs in both clusters. Can we achieve the same thing by issuing a restore request that excludes the HC SMs in one of the two clusters, so that it waits for the HC SMs in the other cluster to restore the database before joining?

sivanov-nuodb replied:
Option 1 will be used by default --- e.g specific backup set or `:group-latest` will restore only HC SMs in one of the clusters. The rest of the SMs will sync.

I decided to include other options in case, the user wants to leverage an accelerated sync use case for all SMs in the database.

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+
+> **NOTE**: A copy of an archive or backup set from another database can't be used to perform _archive seed restore_. Otherwise the Storage Manager process will fail to start with error _Archive "/var/opt/nuodb/archive/nuodb/demo" doesn't match database.  Expected UUID \*\*\*, got \*\*\*._
+
+Identify its archive id using `nuocmd show archives` commands and select it for restore by installing the _restore_ chart with `restore.type="archive"`.
```
Do labels work for seed restore? `nuodocker start sm` injects the label `pod-name`, so specifying `--set restore.labels.pod-name <pod name>` might be a more convenient way to issue the seed restore command (assuming that it works).

sivanov-nuodb replied:
Good idea! That should work. I will test and use that instead.

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+1. Prepare the import source and upload it to a remote location available to NuoDB Storage Manager pods.
+2. Set `database.autoImport` variables during NuoDB database chart installation.
+
+> **NOTE**: For simplicity _nginx_ deployment will be used in this example so that our remote source is available via HTTP in the Kubernetes cluster.
```
_nginx_ is just some arbitrary name for a deployment? Should we choose something more generic sounding like _prod_, _qa_, or _dev_? _qa_ or _dev_ might make sense in this subsection, since that seems like a use-case for importing a database into a fresh domain.

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+
+_Automatic archive restore_ is configured in the `database.autoRestore` section and will execute for every Storage Manager in the database deployment when all of the below conditions are met:
+
+- `database.autoRestore.source` and `database.autoRestore.type` are configured
+- archive metadata is found for the archive in the admin layer or on disk
+- configured restore source is available in the SM pod
+- the archive is detected as corrupted or missing - if the atom files count is less than 20 and catalog atom files count is less than 2
```
Isn't it also a condition that the database has running SMs, otherwise this would be a full database restore? Also, how automated is this, i.e. does it require any user intervention whatsoever?

sivanov-nuodb replied:
There is no check on the database state. I assume that in the general case when no user storage groups are defined --- if we automatically restore the archive from an old backup set, then it won't be selected as leader candidate or the next database cold start.

The `autoRestore` is fully automated and is triggered by SM pod start/restart.

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+
+> **NOTE**: For more information about storage groups, check [Using Table Partitions and Storage Groups](https://doc.nuodb.com/nuodb/latest/database-administration/using-table-partitions-and-storage-groups/)
+
+### Assisted database restore
```
"Assisted" means "assisted by the user". I would just call this "Manual database restore" since that is clearer.

sivanov-nuodb replied:
Okay

adriansuarez commented on docs/BackupAndRestore.md:
```diff
@@ -0,0 +1,594 @@
+
+_Assisted database restore_ is a special case of database restore operation and allows complex restore operations to be executed easier in Kubernetes deployments.
+The _assisted database restore_ operation is initiated by installing the _restore_ chart with `restore.manual="true"` which creates a manual restore request.
+Once the database is restarted all database processes will block waiting for the user to perform the selected archives restore.
```
I wouldn't say "Once the database is restarted" because none of the database processes are actually started. I would make clear what the point is of this mode by including a sentence like

> This mode blocks all database pods before allowing them to form a database, in order to give the user access to the persistent volumes used by an SM.

adriansuarez reviewed (approved):
Excellent document! I left a few comments, but overall, looks very good.

sivanov-nuodb commented:
Thanks, guys for the valuable comments and suggestions!
@acabrele , @adriansuarez I think I addressed most of the comments in case you want to have another look.

sivanov-nuodb commented:
Added user guide on how to use Backup and Restore mechanics with NuoDB Helm Charts in Kubernetes.
